piet-cairo/src/text.rs

```diff
@@ -210,233 +396,198 @@ impl TextLayout for CairoTextLayout {
+        let mut widest_logical_width = 0;
+        let mut widest_whitespaceless_width = 0;
+        let mut iterator = self.pango_layout.get_iter().unwrap();
+        loop {
+            let line = iterator.get_line_readonly().unwrap();
+
+            /*
+             * NOTE: These values are not currently exposed so we need to get them
+             * manually. It's kinda sucky I know
+             *
+             * TODO(ForLoveOfCats): Submit a PR to gtk-rs to expose these values
+             */
+            let (start_offset, end_offset) = unsafe {
+                let raw_line = line.to_glib_none();
+
+                let start_offset = (*raw_line.0).start_index as usize;
+                let length = (*raw_line.0).length as usize;
+
+                (start_offset, start_offset + length)
+            };
+
+            //Pango likes to give us the line range *without* the newline char(s)
+            let end_offset = match self.text.as_bytes()[end_offset..] {
+                [b'\r', b'\n', ..] => end_offset + 2,
+                [b'\r', ..] | [b'\n', ..] => end_offset + 1,
+                _ => end_offset,
+            };
+
+            let logical_rect = line.get_extents().1;
+            if logical_rect.width > widest_logical_width {
+                widest_logical_width = logical_rect.width;
+            }
+
+            let line_text = &self.text[start_offset..end_offset];
+            let trimmed_len = line_text.trim_end().len();
+            let trailing_whitespace = line_text[trimmed_len..].len();
+
+            widest_whitespaceless_width = {
+                let start_x = line.index_to_x(trimmed_len as i32, false);
+                let end_x = line.index_to_x(line_text.len() as i32, true);
+                let whitespace_width = end_x - start_x;
+                let whitespaceless_width = logical_rect.width - whitespace_width;
+                widest_whitespaceless_width.max(whitespaceless_width)
+            };
+
+            line_metrics.push(LineMetric {
+                start_offset,
+                end_offset,
+                trailing_whitespace,
+                baseline: (iterator.get_baseline() as f64 / PANGO_SCALE) - y_offset,
+                height: logical_rect.height as f64 / PANGO_SCALE,
+                y_offset,
+            });
+            y_offset += logical_rect.height as f64 / PANGO_SCALE;
+
+            if !iterator.next_line() {
+                break;
+            }
+        }
 
-    let text_len = text.len();
+        //NOTE: Pango appears to always give us at least one line even with empty input
+        self.line_metrics = line_metrics;
 
-    if text_position == 0 {
-        return 0.0;
-    }
+        let ink_extent = self.pango_layout.get_extents().0;
+        let logical_extent = self.pango_layout.get_extents().1;
+        self.size = Size::new(
+            widest_whitespaceless_width as f64 / PANGO_SCALE,
+            logical_extent.height as f64 / PANGO_SCALE,
+        );
```

Comment:
you can write this as `Size::new(x, y) / PANGO_SCALE`, which might be slightly cleaner?

Reply:
I personally find this more clear but more importantly written the other way it would still need to be split over multiple lines which makes the division a bit weird (this is how rustfmt writes it)
```rust
Size::new(
    widest_whitespaceless_width as f64,
    logical_extent.height as f64,
) / PANGO_SCALE;
```